drivers/at86rf2xx/at86rf2xx.c

```diff
@@ -32,59 +35,215 @@
+        at86rf2xx_setup((at86rf2xx_t *)devs);
+        devs++;
+        params++;
+    }
+#else
+    (void)devs;
+    (void)params;
+    (void)num;
+#endif
+}
+
+void at86rfa1_setup(at86rfa1_t *dev)
+{
+#if IS_USED(MODULE_AT86RFA1)
+    dev->base.dev_type = AT86RF2XX_DEV_TYPE_AT86RFA1;
+    at86rf2xx_setup((at86rf2xx_t *)dev);
+#else
+    (void)dev;
+#endif
+}
+
+void at86rfr2_setup(at86rfr2_t *dev)
+{
+#if IS_USED(MODULE_AT86RFR2)
+    dev->base.dev_type = AT86RF2XX_DEV_TYPE_AT86RFR2;
+    at86rf2xx_setup((at86rf2xx_t *)dev);
+#else
+    (void)dev;
+#endif
+}
+
+size_t at86rf2xx_get_size(const at86rf2xx_t *dev)
+{
+    switch (dev->base.dev_type) {
+        default: return sizeof(at86rf2xx_t);
+#if IS_USED(MODULE_AT86RF212B)
+        case AT86RF2XX_DEV_TYPE_AT86RF212B: {
+            return sizeof(at86rf212b_t);
+        }
+#endif
+#if IS_USED(MODULE_AT86RF231)
+        case AT86RF2XX_DEV_TYPE_AT86RF231: {
+            return sizeof(at86rf231_t);
+        }
+#endif
+#if IS_USED(MODULE_AT86RF232)
+        case AT86RF2XX_DEV_TYPE_AT86RF232: {
+            return sizeof(at86rf232_t);
+        }
+#endif
+#if IS_USED(MODULE_AT86RF233)
+        case AT86RF2XX_DEV_TYPE_AT86RF233: {
+            return sizeof(at86rf233_t);
+        }
+#endif
+#if IS_USED(MODULE_AT86RFA1)
+        case AT86RF2XX_DEV_TYPE_AT86RFA1: {
+            return sizeof(at86rfa1_t);
+        }
+#endif
+#if IS_USED(MODULE_AT86RFR2)
+        case AT86RF2XX_DEV_TYPE_AT86RFR2: {
+            return sizeof(at86rfr2_t);
+        }
+#endif
+    }
+}
+
+static void at86rf2xx_disable_clock_output(at86rf2xx_t *dev)
+{
+    switch (dev->base.dev_type) {
+        default:;
+            uint8_t tmp = at86rf2xx_reg_read(dev, AT86RF2XX_REG__TRX_CTRL_0);
+            tmp &= ~(AT86RF2XX_TRX_CTRL_0_MASK__CLKM_CTRL);
+            tmp &= ~(AT86RF2XX_TRX_CTRL_0_MASK__CLKM_SHA_SEL);
+            tmp |= (AT86RF2XX_TRX_CTRL_0_CLKM_CTRL__OFF);
+            at86rf2xx_reg_write(dev, AT86RF2XX_REG__TRX_CTRL_0, tmp);
+            break;
+#if IS_USED(MODULE_AT86RFA1)
+        case AT86RF2XX_DEV_TYPE_AT86RFA1: break;
+#endif
+#if IS_USED(MODULE_AT86RFR2)
+        case AT86RF2XX_DEV_TYPE_AT86RFR2: break;
 #endif
+    }
+}
+
+static void at86rf2xx_enable_smart_idle(at86rf2xx_t *dev)
+{
+    switch (dev->base.dev_type) {
+        default:
+            (void)dev;
+            break;
+#if IS_USED(MODULE_AT86RF233)
+        case AT86RF2XX_DEV_TYPE_AT86RF233: {
+            uint8_t tmp = at86rf2xx_reg_read(dev, AT86RF2XX_REG__TRX_RPC);
+            tmp |= (AT86RF2XX_TRX_RPC_MASK__RX_RPC_EN |
+                    AT86RF2XX_TRX_RPC_MASK__PDT_RPC_EN |
+                    AT86RF2XX_TRX_RPC_MASK__PLL_RPC_EN |
+                    AT86RF2XX_TRX_RPC_MASK__XAH_TX_RPC_EN |
+                    AT86RF2XX_TRX_RPC_MASK__IPAN_RPC_EN);
+            at86rf2xx_reg_write(dev, AT86RF2XX_REG__TRX_RPC, tmp);
+            at86rf2xx_set_rxsensitivity(dev, AT86RF233_RSSI_BASE_VAL);
+            break;
+        }
+#endif
+    }
```

Comment:
I think the RFA1 and the RFR2 also support smart idle and this was recently merged upstream.

Reply:
Only RFR2 does.

Reply:
Ditto. With device specific functions (`at86rf233_enable_smart_idle`) we get rid of the switch-case.

Reply:
Here, a device specific function makes sens to me.
However, the calling function must check if the device type

But the implementation of `at86rf233_enable_smart_idle()` and `at86rfr2_enable_smart idle()` would be the same.
So I think the best would be to take the core of `at86rf2xx_enable_smart_idle()`, keep the name,  and make static inline `at86rf233_enable_smart_idle()` and `at86rfr2_enable_smart idle()`
static inline wrappers and the calling function (`at86rf2xx_reset()`) performs the switch case type checking.

Reply:
Yes, that makes sense!

Reply:
> However, the calling function must check the device type

Why?
We could move that logic on top. I would assume the user of the raw driver knows which device is being used (AT86RF231, AT86RF233, etc). I would define the driver with variants specific prefixes (most of them can be just static inline wrappers to a common function) and let the layer on top (e.g netdev) device which function should be called.
Does this makes sense?